core-rust/state-manager/src/staging/result.rs

```diff
@@ -373,29 +376,98 @@ impl ProcessedCommitResult {
+    /// `{added: #1#, removed: #1#}` may be encountered here (and it simply means that resources
+    /// were only moved around different vaults of the same global root entity).
+    pub global_balance_changes: IndexMap<GlobalAddress, IndexMap<ResourceAddress, BalanceChange>>,
+
+    /// A resultant balances of fungible resources held by global accounts.
+    /// Note: aggregating a resultant balance of *any* global root entity is, in principle, not
+    /// possible using only the "vault balance changes" input (because there may be vaults whose
+    /// balance was not changed). We "aggregate" it for accounts using an assumption that an account
+    /// has at most one vault for each resource.
+    pub resultant_fungible_account_balances:
+        IndexMap<GlobalAddress, IndexMap<ResourceAddress, Decimal>>,
+}
+
+impl GlobalBalanceSummary {
+    /// Computes a [`GlobalBalanceSummary`] from the given vault balance changes.
+    /// This uses the ancestry information from the given [`SubstateNodeAncestryStore`] to resolve
+    /// the owning global entity, and parses the substate changes to collect the resulting balances
+    /// of fungible vaults.
+    pub fn compute_from<S: SubstateNodeAncestryStore>(
         store: &S,
         vault_balance_changes: &IndexMap<NodeId, (ResourceAddress, BalanceChange)>,
-    ) -> IndexMap<GlobalAddress, IndexMap<ResourceAddress, BalanceChange>> {
+        substate_changes: &BySubstate<ChangeAction>,
+    ) -> Self {
         let ancestries = store.batch_get_ancestry(vault_balance_changes.keys());
         let mut global_balance_changes = index_map_new();
+        let mut resultant_fungible_account_balances = index_map_new();
         for (vault_entry, ancestry) in vault_balance_changes.iter().zip(ancestries) {
             let (vault_id, (resource_address, balance_change)) = vault_entry;
             let root = ancestry
                 .map(|ancestry| ancestry.root.0)
                 .unwrap_or_else(|| *vault_id);
-            let Ok(global_ancestor_address) = GlobalAddress::try_from(root) else {
+            let Ok(root_address) = GlobalAddress::try_from(root) else {
                 panic!("root {:?} resolved for vault {:?} is not global", root, vault_id);
             };
+
+            // Aggregate (i.e. sum) balance changes for every global root entity.
             global_balance_changes
-                .entry(global_ancestor_address)
+                .entry(root_address)
                 .or_insert_with(index_map_new::<ResourceAddress, BalanceChange>)
                 .entry(*resource_address)
                 .and_modify(|existing| *existing = existing.clone() + balance_change.clone())
                 .or_insert_with(|| balance_change.clone());
+
+            // Collect (i.e. not sum) resultant balances for fungible resources of global accounts.
+            if vault_id.is_internal_fungible_vault() && root_address.is_account() {
```

Comment:
Can we also check the ancestry that the vault has a parent substate in the `AccountPartitionOffset::ResourceVaultKeyValue` partition?

Ie that `ancestry.unwrap().parent.1 == AccountPartitionOffset::ResourceVaultKeyValue.as_main_partition()`?

This will protect us against the possibility of Account getting other vaults in future, or EG the royalty vault coming back(!)

Reply:
I'll make these changes and merge